File: Data Structures & Algorithms/surrounded-regions/submission-2.java
class Solution {
    public void dfs(int r, int c, char[][] board) {
        if (r < 0 || r >= board.length || c < 0 || c >= board[0].length
        || board[r][c] != 'O') {
            return;
        }

        board[r][c] = 'T';

        dfs(r + 1, c, board);
        dfs(r - 1, c, board);
        dfs(r, c + 1, board);
        dfs(r, c - 1, board);

        return; 
    }

    public void solve(char[][] board) {
        int ROWS = board.length;
        int COLS = board[0].length;

        for (int r = 0; r < ROWS; r++) {
            dfs(r, 0, board);
            dfs(r, COLS - 1, board);
        }

        for (int c = 0; c < COLS; c++) {
            dfs(0, c, board);
            dfs(ROWS - 1, c, board);
        }

        for (int r = 0; r < ROWS; r++) {
            for (int c = 0; c < COLS; c++) {
                if (board[r][c] == 'O') {
                    board[r][c] = 'X';
                } else if (board[r][c] == 'T') {
                    board[r][c] = 'O';
                }
            }
        }
    }
}

File: Data Structures & Algorithms/search-for-word/submission-2.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        ROWS = len(board)
        COLS = len(board[0])

        directions = [(0, 1), (1, 0), (-1, 0), (0, -1)]
        def dfs(r, c, i, path):
            if i == len(word):
                return True

            if r < 0 or r == ROWS or c < 0 or c == COLS:
                return False

            if board[r][c] != word[i]:
                return False
            
            path += board[r][c]
            # print(path)

            for direction in directions:
                if dfs(r + direction[0], c + direction[1], i + 1, path[:]):
                    return True
            
            return False

        for r in range(ROWS):
            for c in range(COLS):
                if board[r][c] == word[0]:
                    if (dfs(r, c, 0, '')):
                        return True
        
        return False

